src/Knp/Menu/Matcher/Voter/RequestStackRouteVoter.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace Knp\Menu\Matcher\Voter;
+
+use Knp\Menu\ItemInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Voter based on the route
+ */
+class RequestStackRouteVoter implements VoterInterface
+{
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @param RequestStack $requestStack
+     */
+    public function __construct(RequestStack $requestStack)
+    {
+        $this->requestStack = $requestStack;
+    }
+
+    public function matchItem(ItemInterface $item)
+    {
+        if (null === $request = $this->requestStack->getCurrentRequest()) {
```

Comment:
should we have a config option whether to use current or master request? or two different classes for the 2 cases. the current implementation with the request listener only updates voters with the master request... so i would default to use the master request.